api/grpcserver/mesh_service.go

```diff
@@ -529,28 +529,51 @@ func convertLayerStatus(in int) pb.Layer_LayerStatus {
 
 func (s MeshService) EpochStream(req *pb.EpochStreamRequest, stream pb.MeshService_EpochStreamServer) error {
 	epoch := types.EpochID(req.Epoch)
-	var total, mal int
-	if err := s.cdb.IterateEpochATXHeaders(epoch+1, func(header *types.ActivationTxHeader) error {
-		total++
-		select {
-		case <-stream.Context().Done():
-			return nil
-		default:
-			malicious, err := s.cdb.IsMalicious(header.NodeID)
-			if err != nil {
-				return err
-			}
-			if malicious {
-				mal++
-				return nil
+	var (
+		dberr      error
+		total, mal int
+	)
+	query := `SELECT
+		atxs.id,
+		atxs.pubkey,
+		iif(i.proof is null, 0, 1) as malicious
+		FROM atxs left join identities i on atxs.pubkey = i.pubkey WHERE atxs.epoch = $1 `
+
+	_, err := s.cdb.Exec(query,
```

Comment:
It would probably be best to have all the code that deals with the database under `sql/`
In this case it should be possible to use a per-row callback fn